File: haskell-playground/haskell-solutions/src/MySolutions/Numbers/LargestNumberUnderNDivisibleByAGivenNumber.hs
module MySolutions.Numbers.LargestNumberUnderNDivisibleByAGivenNumber (largestDivisible) where

largestDivisible :: Integral a => a -> a -> a
largestDivisible n d = head (filter p [n,n-1..])
    where p x = x `mod` d == 0

-- We do not have to start with a finite list for our starting set.
-- laziness is in action here.
-- The evaluation stops when the first adequate solution is found.

-- tests
testLargestDivisible01 = largestDivisible 100000 3829
-- 99554

File: java-playground/my-implementations/algorithms-and-datastructures/src/main/java/com/my/company/numbers/LargestNumberUnderNDivisibleByAGivenNumber.java
package com.my.company.numbers;

public class LargestNumberUnderNDivisibleByAGivenNumber {

    public int getLargestNumberUnderNDivisibleByD(int N, int divisor) {

        if (N <= 0 || divisor <= 0) {
            return 0;
        }

        for (int i = N - 1; i > 0; i--) {
            if (i % divisor == 0) {
                return i;
            }
        }

        return 0;
    }
}
